openedx/core/djangoapps/content/learning_sequences/api/data.py

```diff
@@ -0,0 +1,165 @@
+"""
+Note: we're using old-style syntax for attrs because we need to support Python
+3.5, but we can move to the PEP-526 style once we move to Python 3.6+.
+
+Note: These attr classes are only allowed to do validation that is entirely
```

Comment:
I'm still reading the PR so maybe I'll just see this, but how is the spirit of these different than a serializer?

Reply:
Or rather, I understand that a serializer is in preparation for sending things over the wire in a response, but I wonder how people will know where a particular change to the data model should happen, here, or closer to the response formatting.

Reply:
I think most changes should be at this layer, with the response formatting being as thin and dumb as possible. The goal would be to make the process-internal API usable and fully featured, and to not have REST-only API functionality. So in this scenario, the changes you'd need when you altered the data model would be...

**Data classes (almost always)**: The data classes here are the only complex objects accepted by or passed back by this app's public API. Any changes you want to make in the data model probably require changes here.

**Serializers (when it needs to go to the browser)**: The serializers are purely a convenience of the REST API view layer. The job of the view layer is to translate back and forth between user inputs and these data classes. New fields in data classes would be echoed out here, though they may be formatted/arranged very differently to make it more convenient for JS code to work with. It's possible that new things would be added to the in-process API that are never reflected in Serializers/REST because they are only useful to different parts of the LMS that need some bit of course outline information to run checks against.

**Models (sometimes)**: New data we need to persist will of course require model changes to store them, but there are data classes like `UserCourseOutlineDetailsData` which will be composed mostly of data that's already accessible via different APIs/models already in edx-platform.

Reply:
Ahh - so the data classes are our way of ensuring that callers of the apps public API aren't abusing/misusing the ORM after they receive a model?  (I recall a discussion of this at some point, maybe at an arch tea time)

This all makes sense.

Reply:
Yes! Also, a way to make it easier for testing (since these are dumb structs that are easy to create) and mock (for apps that call into this).

Reply:
So for instance, I was really tempted to attach the OutlineProcessors to the UserCourseOutline in case there was extra queries you might want to make on specific processors (e.g. schedule details). But that would have given a really big surface area to mock out as part of the public interface (who knows what processors implement what non-inherited methods?).